File: packages/typescript/tests/system/tabs.test.ts
import { SwitchToNextTabTool, SwitchToPreviousTabTool } from "alumnium";
import { describe } from "vitest";
import { baseIt } from "./helpers.js";

describe("Tabs", () => {
  const it = baseIt.override("setup", async ({ setup, skip }) => {
    return async (options) => {
      const result = await setup(options);
      const { driverType } = result;

      if (driverType === "appium")
        skip("Tabs functionality is not implemented in Appium yet");

      return result;
    };
  });

  it("switching tabs", async ({ expect, setup }) => {
    const { al, $ } = await setup({
      extraTools: [SwitchToNextTabTool, SwitchToPreviousTabTool],
    });

    await $.navigate("multi_tab_page.html");

    await al.do("click on 'Open New Tab' button");
    expect(await al.get("current page URL")).toBe("about:blank");

    await al.do("switch to previous browser tab");
    expect(await al.get("header text")).toBe("Multi-Tab Test Page");

    await al.do("switch to next browser tab");
    expect(await al.get("current page URL")).toBe("about:blank");

    await al.do("switch to next browser tab");
    expect(await al.get("header text")).toBe("Multi-Tab Test Page");

    await al.do("switch to previous browser tab");
    expect(await al.get("current page URL")).toBe("about:blank");
  });
});

File: packages/java/src/test/java/ai/alumnium/system/TabsTest.java
package ai.alumnium.system;

import ai.alumnium.tool.SwitchToNextTabTool;
import ai.alumnium.tool.SwitchToPreviousTabTool;
import java.io.File;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class TabsTest extends BaseTest {

  static {
    extraTools = List.of(SwitchToNextTabTool.class, SwitchToPreviousTabTool.class);
  }

  private static final String MULTI_TAB_URL =
      new File("../python/examples/support/pages/multi_tab_page.html").toURI().toString();

  @Test
  void testSwitchingTabs() {
    navigate(MULTI_TAB_URL);

    al.act("click on 'Open New Tab' button");
    Assertions.assertEquals("about:blank", al.get("current page URL").toObject());

    al.act("switch to previous browser tab");
    Assertions.assertEquals("Multi-Tab Test Page", al.get("header text").toObject());

    al.act("switch to next browser tab");
    Assertions.assertEquals("about:blank", al.get("current page URL").toObject());

    al.act("switch to next browser tab");
    Assertions.assertEquals("Multi-Tab Test Page", al.get("header text").toObject());

    al.act("switch to previous browser tab");
    Assertions.assertEquals("about:blank", al.get("current page URL").toObject());
  }
}
